### ISSUE TITLE
Remount component if componentWillMount\componentDidMount changed

### ISSUE THREAD

**Issue by Grimones** [gaearon/react-hot-loader#771]
Hi!

After reading theKasheys [post](https://habrahabr.ru/post/346142/) (russian language) about RHL v4 i have finally understood  how new rhl works (at least i think so). And i have an idea about what you can do with changing `componentWillMount`\\`componentDidMount`.

Now you must reload the page when you change `componentWillMount`\\`componentDidMount`.

If RHL can detect that `componentWillMount`\\`componentDidMount` has changed maybe it will make sense to add in configuration something like

```jsx
import { setConfig } from 'react-hot-loader';
setConfig({ logLevel: 'debug', remountComponentOn: ['cwm', cdm'] });
```
where `cwm` and `cdm` is abbreaviation for` componentWillMount` and `componentDidMount`

And if you have changed the most inner component, it means that parents can preserve their state and remount only the child where `componentWillMount`\\`componentDidMount` has changed. I believe that it will be better than reloading the whole page. At least some state of the app will be saved. 

Does this make sense for you? I guess it is real to implement.

**Comment by theKashey** [comment]
Yep, I thought about it. The problem is - you cant always detect a change inside cwm/cdm. For example
```js
import {getInterval, setIntervalFunction} from './utils';

class Ticker extends Components {
  tick = () => this.setState({counter: this.state.counter+1});
  
  componentDidMount() {
    setIntervalFunction(this.tick, getInterval())
  }
}
```
RHL could detect the change inside cdm, but could not detect the change inside tick, setIntervalFunction or getInterval. And I expect most of the changes will be `external`. 
There are so many side effects, could be done in lifecycle methods. We just dont know how to detect them. Yet.

**Comment by Grimones** [comment]
Thanks for the explanation. Considering the progress with v4 i believe that you will find out how to detect that changes :)
